src/types.rs

```diff
@@ -50,13 +52,36 @@ pub struct LakeConfig {
     ///         .expect("Failed to build LakeConfig");
     /// # }
     /// ```
+    ///
+    /// This field is mutually exclusive with [LakeConfigBuilder::s3_client].
     #[builder(setter(strip_option), default)]
     pub s3_config: Option<aws_sdk_s3::config::Config>,
+    /// Provide a custom S3 client which implements the s3_fetchers::S3Client trait. This is useful
+    /// if you need more control over the requests made to S3, e.g. you want to add cache.
+    ///
+    /// This field is mutually exclusive with [LakeConfigBuilder::s3_config].
+    #[builder(setter(strip_option, custom), default)]
+    pub(crate) s3_client: Option<Box<dyn S3Client>>,
     #[builder(default = "100")]
     pub(crate) blocks_preload_pool_size: usize,
 }
 
 impl LakeConfigBuilder {
+    fn validate(&self) -> Result<(), String> {
+        if self.s3_config.is_some() && self.s3_client.is_some() {
+            return Err("Cannot provide both s3_config and s3_client".to_string());
+        }
+
+        Ok(())
+    }
+
+    pub fn s3_client<T: S3Client + 'static>(self, s3_client: T) -> Self {
```

Comment:
Custom setter which allows passing `S3Client` directly, rather than requiring user to wrap in `Box`.